Support Z-Wave JS dimming lights using color intensity

AlCalzone opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

In Z-Wave, colored lights exist that can only be dimmed through the color channels, but not a separate brightness channel. An example is the MATRIX ZDB5100 Wall Switch, where the LED indicators can be turned on and off, and color changed: [MATRIX_ZDB5100_User_Manual_1_01-EN-2.pdf](https://github.com/user-attachments/files/16389728/MATRIX_ZDB5100_User_Manual_1_01-EN-2.pdf). Another is the HomeSeer HSM200 v1, which only implements Color Switch CC, no Binary or Multilevel Switch.

This was hard-coded into discovery before for these two devices, but dimming through the brightness controls didn't work. Certification also tests the combination of Binary Switch + Color Switch.

This PR adds a discovery schema and updates the handling of the Z-Wave Light class to be able to control those kinds of devices.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

fixes: https://github.com/home-assistant/core/issues/120139
fixes: https://github.com/zwave-js/certification-backlog/issues/19

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

AlCalzone committed:
Z-Wave JS: support non-dimmable color lights

AlCalzone commented:
The test `test_black_is_off_zdb5100` in `tests/components/zwave_js/test_light.py` is failing, because it expects `turn_on` to set the color. With this change, a Binary Switch On command is sent instead.

IMO this makes sense, but the manual of the device states that
a) Color Switch CC alone is enough to control the LED indicators, which I assume is the reason for the `black_is_off` attribute
b) The default behavior of Binary Switch CC (toggle LEDs) can be changed to control both the LEDs and the relays, so this change may be unintended for this particular device after all.

MartinHjelmare commented on homeassistant/components/zwave_js/light.py:
```diff
@@ -130,14 +130,17 @@ def __init__(
+            )
+            self._supports_dimming = False
+        elif self.info.primary_value.command_class == CommandClass.BASIC:
+            # If the command class is Basic, we must geenerate a name that includes
```
```suggestion
            # If the command class is Basic, we must generate a name that includes
```

MartinHjelmare reviewed (commented):
The changes you made to the target values in the ZwaveLight class affects the ZwaveBlackIsOffLight too since it inherits the ZwaveLight class. The test entity still belongs to the same discovery schema.

MartinHjelmare commented:
Besides solving these problems we should add new tests for the new discovery schema.

AlCalzone commented:
I've tried to debug the test, but the `turn_on` method of `ZwaveBlackIsOffLight` is no longer called. When I revert the changes, it is.

AlCalzone commented:
I tend to agree with your comment there:
> I'm not sure HS is the right color mode for Z-Wave devices in general

...given that Z-Wave operates on individual color components

AlCalzone commented:
I think the current implementation is not entirely correct, given the specification requirements:

`ZwaveLight` follows the above recommendation of using the highest possible color component level by forcing an RGB light to HS and discarding the lightness component.

`ZwaveBlackIsOffLight` seems to be an escape hatch, but it does not fully support dimming. If I understand the code correctly, it assumes full brightness and realizes an OFF state by setting all RGB channels to 0. IMO there are two problems with that:
1. No dimming, as HS has no intermediate levels between full lightness and black, and the brightness property always returns 255
2. `turn_off` sets the brightness to 0, but `turn_on` does not and assumes that setting the color to a non-zero value turns the device on. Per the specifications, this assumption is not correct though:

> The Color Switch Command Class commands MUST NOT affect the brightness level controlled by
Basic or Binary/Multilevel Switch Command Classes.

I think we should change `ZwaveBlackIsOffLight` to use the RGB color mode instead and have it's `turn_on` method set the brightness just like `ZwaveLight` does.

AlCalzone committed:
remove black_is_off light, support on/off/color

AlCalzone committed:
fix: tests for on/off light

AlCalzone committed:
fix: typo

AlCalzone commented:
I've reworked this a bit:
Removed `ZwaveBlackIsOffLight` for the aforementioned reasons. Instead I added support for mixing brightness and color channels for non-dimmable lights in the `ZwaveLight` class:
- When only brightness is set, the previous color is read and set again, scaled with the brightness. Turn binary switch on or off
- When only the color is set, the current brightness is used to scale the color before setting it. This prevents the brightness from jumping back to 100% on every color change
- The current brightness is inferred from the highest color channel, e.g. RGB (220, 0, 50) means brightness = 220/255. (0,0,0) means off.

AlCalzone commented on tests/components/zwave_js/test_light.py:
```diff
@@ -597,125 +615,67 @@ async def test_black_is_off(
     client.async_send_command.reset_mock()
 
-    # Force the light to turn off
+    # Turn the light on and set it to green (50% brightness)
```
Is there an easier way to reset the call args, but preserve the current mock state?
This seems unnecessarily repetitive.

MartinHjelmare replied:
If you just want to test the service call you can just remove the receive of the event. Is that what you mean?

AlCalzone replied:
The sent commands depend on the current state (brightness / color) of the light, so having the correct state before calling the service is necessary.

These assertions depend on how many calls were done before, which introduces coupling between the sub-tests
```py
    assert len(client.async_send_command.call_args_list) == 1
    args = client.async_send_command.call_args_list[0][0][0]
```

so I'm looking for a way to reset the call history, but not the entire mock state.

MartinHjelmare replied:
You can use `reset_mock` on the mock attribute that you want to reset.

AlCalzone replied:
Ok so just `client.async_send_command.call_args_list.reset_mock()`?

MartinHjelmare replied:
No, `client.async_send_command.reset_mock()`.

AlCalzone committed:
remove commented out old test code

AlCalzone committed:
add test for off and on

AlCalzone commented:
What does it support? I have one, but I think it's a later model. Works fine as-is.

AlCalzone commented:
I'll double-check that tomorrow.

AlCalzone committed:
support colored lights without separate brightness control

AlCalzone committed:
add test for color-only light

AlCalzone committed:
refactor: extract color only light

AlCalzone committed:
fix: preserve color when changing brightness

AlCalzone commented:
- Tests for the Color Switch CC only light restored
- Refactored into a different class for color only lights again (yeah, I know... 😅)

Here's a video - Node 2 is Color only, Node 3 is Color + Binary
[Screencast from 31.07.2024 16:43:01.webm](https://github.com/user-attachments/assets/7562926a-fd07-4dd1-9bb1-f0d8b494d911)

MartinHjelmare commented:
Please make sure that all changed lines are covered by tests.

AlCalzone commented:
How can I check this locally?

AlCalzone committed:
extend tests

AlCalzone committed:
refactor again

AlCalzone committed:
refactor scale check

AlCalzone committed:
refactor: remove impossible check

MartinHjelmare commented on homeassistant/components/zwave_js/light.py:
```diff
@@ -499,61 +542,137 @@ def __init__(
+
+        # Brightness is encoded in the color channels by scaling them lower than 255
+        color_values = [v.value for v in self._get_color_values() if v is not None]
+        max_value = max([v for v in color_values if v is not None])
```
This will error if the list is empty.

```py
Python 3.12.4 (main, Jun  8 2024, 18:29:57) [GCC 9.4.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.22.2 -- An enhanced Interactive Python. Type '?' for help.

In [1]: max([])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[1], line 1
----> 1 max([])

ValueError: max() iterable argument is empty
```

If we know that this light always supports red, green and blue colors, ie we don't have to worry about an empty list here, it's weird that we check `self._supports_color` in this class.

AlCalzone replied:
The modified discovery scheme creates an instance of this class for
a) devices that only have a current color, but no switching capabilities
b) devices that have a binary switch and a current color

So there should always be at least one color component, otherwise the scheme wouldn't match.

> it's weird that we check self._supports_color in this class.

Due to the discovery scheme, it might be that we end up with a white-spectrum light that cannot be dimmed, but turned on and off. I'm not sure how to match lights that have exactly RGB support, but nothing else, since the discovery scheme seems to match when ANY of the specified values are there, not ALL.

Should the initial `if` in `async_turn_on` and `brightness` should defer to the parent class if there is no RGB support? That way we should avoid this problem altogether.

MartinHjelmare commented on homeassistant/components/zwave_js/light.py:
```diff
@@ -499,61 +542,137 @@ def __init__(
+        # Brightness is encoded in the color channels by scaling them lower than 255
+        color_values = [v.value for v in self._get_color_values() if v is not None]
+        max_value = max([v for v in color_values if v is not None])
+        return max_value if max_value is not None else 0
```
`max_value` can't be `None` with the filtering above it seems.

AlCalzone commented on homeassistant/components/zwave_js/light.py:
```diff
@@ -499,61 +542,137 @@ def __init__(
+
+        # Remember last color and brightness to restore it when turning on
+        self._last_brightness = self.brightness
+        current_color = self.get_zwave_value(
```
I guess we can cache it in `_calculate_color_values`, but I was under the impression that this is just a simple lookup.

MartinHjelmare replied:
If it's the same instance there's no point in looking it up more than once.

AlCalzone committed:
review feedback

AlCalzone commented on homeassistant/components/zwave_js/light.py:
```diff
@@ -499,61 +542,137 @@ def __init__(
+            ):
+                # Changed brightness from 0 to >0
+                old_brightness = max(self._last_on_color.values())
+                new_scale = brightness / old_brightness
```
I had to do this because `scale` is considered `float | None`, even after the definitive assignment here. The suggestion causes an error on the line where `scale` is multiplied.

Maybe there's a different trick I'm not aware of, like TypeScript's non-null assertion?

MartinHjelmare replied:
Ok.

AlCalzone committed:
review feedback

MartinHjelmare committed:
Merge branch 'dev' into zwave-on-off-color

MartinHjelmare reviewed (approved):
Looks good!
